observatory-dags/observatory/dags/workflows/onix_workflow.py

```diff
@@ -0,0 +1,314 @@
+)
+
+
+class OnixWorkflow(Telescope):
+    """This workflow telescope:
+    1. [Not implemented] Creates an ISBN13-> internal identifier lookup table.
+    2. Creates an ISBN13 -> WorkID lookup table.
+      a. Aggregates Product records into Work clusters.
+      b. Writes the lookup table to BigQuery.
+      c. Writes an error table to BigQuery.
+    3. Create an ISBN13 -> Work Family ID lookup table.  Clusters editions together.
+      a. Aggregate Works into Work Families.
+      b. Writes the lookup table to BigQuery.
+    """
+
+    DAG_ID_PREFIX = "onix_workflow"
+
+    def __init__(
+        self,
+        *,
+        org_name: str,
+        gcp_project_id: str,
+        gcp_bucket_name: str,
+        telescope_sensor: ExternalTaskSensor,
+        dag_id: Optional[str] = None,
+        start_date: Optional[datetime] = datetime(2021, 3, 28),
+        schedule_interval: Optional[str] = "@weekly",
+        catchup: Optional[bool] = False,
+        bq_client: Optional[Union[None, Client]] = None,
+    ):
+        """Initialises the workflow object.
+        :param org_name: Organisation name.
+        :param gcp_project_id: Project ID in GCP.
+        :param gcp_bucket_name: GCP bucket name to store files.
+        :param telescope_sensor: Telescope sensor to monitor.
+        :param dag_id: DAG ID.
+        :param start_date: Start date of the DAG.
+        :param schedule_interval: Scheduled interval for running the DAG.
+        :param catchup: Whether to catch up missed DAG runs.
+        """
+
+        self.dag_id = dag_id
+        if dag_id is None:
+            self.dag_id = make_dag_id(self.DAG_ID_PREFIX, org_name)
+
+        # Set the Bigquery client object
+        self.client = bq_client
+        if bq_client is None:
+            self.client = Client()
+
+        # GCP parameters
+        self.workflow_dataset_id = "onix_workflow"
+        self.project_id = gcp_project_id
+        self.onix_dataset_id = "onix"
+        self.dataset_location = "us"
+        self.dataset_description = "ONIX workflow tables"
+
+        # ONIX release info
+        table_name, onix_release_date_str = self.get_latest_table_id()
+        self.onix_table_id = table_name
+        self.release_date = pendulum.parse(onix_release_date_str)
+        self.bucket_name = gcp_bucket_name
+        self.blob_path = os.path.join("onix_workflow", onix_release_date_str)
+
+        # Prepare filesystem
+        self.local_table_dir = self.blob_path
+        self.workslookup_filename = os.path.join(self.local_table_dir, "onix_workid_isbn.jsonl.gz")
+        self.workslookup_errors_filename = os.path.join(self.local_table_dir, "onix_workid_isbn_errors.jsonl.gz")
+        self.worksfamilylookup_filename = os.path.join(self.local_table_dir, "onix_workfamilyid_isbn.jsonl.gz")
+
+        Path(".", self.local_table_dir).mkdir(exist_ok=True, parents=True)
+
+        # Work (family) aggregation results
+        self.works = list()
+        self.works_families = list()
```

Comment:
It would be good to put these into a release object similar to the other DAGs and create the release object as part of an Airflow task.

The reason is that, the OnixWorkflow is constructed, and therefore the `release_date` is fetched, when the DAG file is loaded, not after waiting for the ONIX telescope to finish (via the external task sensor).